paasta_tools/cli/cmds/mark_for_deployment.py

```diff
@@ -347,30 +432,33 @@ def wait_for_deployment(service, deploy_group, git_sha, soa_dir, timeout):
         raise NoInstancesFound
     paasta_print("Waiting for deployment of {0} for '{1}' complete..."
                  .format(git_sha, deploy_group))
-    for cluster in cluster_map.values():
-        cluster['deployed'] = []
-    total_instances = sum([len(v["instances"]) for v in cluster_map.values()])
+
+    total_instances = 0
+    for cluster in cluster_map:
+        cluster_map[cluster]['deployed'] = 0
+        cluster_map[cluster]['queue'] = Queue()
+        cluster_map[cluster]['num_instances'] = len(cluster_map[cluster]['instances'])
+        total_instances += cluster_map[cluster]['num_instances']
+        for i in cluster_map[cluster]['instances']:
+            cluster_map[cluster]['queue'].put(i)
     deadline = time.time() + timeout
+    green_light = Event()
+    green_light.set()
 
     with progressbar.ProgressBar(maxval=total_instances) as bar:
         while time.time() < deadline:
-            for cluster, instances in cluster_map.items():
-                if len(cluster_map[cluster]['deployed']) != len(cluster_map[cluster]['instances']):
-                    cluster_map[cluster]['deployed'] += instances_deployed(
-                        cluster=cluster,
-                        service=service,
-                        instances=[i for i in instances['instances'] if i
-                                   not in cluster_map[cluster]['deployed']],
-                        git_sha=git_sha)
-                    if len(cluster_map[cluster]['deployed']) == len(cluster_map[cluster]['instances']):
-                        instance_csv = ", ".join(cluster_map[cluster]['instances'])
-                        paasta_print("Deploy to %s complete! (instances: %s)" % (cluster, instance_csv))
-                bar.update(sum([len(v["deployed"]) for v in cluster_map.values()]))
-            if all([len(cluster['deployed']) == len(cluster["instances"]) for cluster in cluster_map.values()]):
+            _query_clusters(cluster_map, service, git_sha, green_light)
+            if not green_light.is_set():
+                raise KeyboardInterrupt
```

Comment:
I'm afraid I disagree. This is exactly the correct exception.

This event isn't set only if KeyboardInterrupt was received by a running thread.
https://github.com/Yelp/paasta/pull/972/files#diff-c18610688c26664370ccca67b0151768R424
I'm just re-raising the same exception in the main thread to follow the same code path as it was before.